docs/proposals/env-vars.md

```diff
@@ -0,0 +1,316 @@
+	// will be reported as an event when the container is starting. When a key exists in multiple
+	// sources, the value associated with the last source will take precedence.
+	// Values defined by an Env with a duplicate key will take precedence.
+	// Cannot be updated.
+	// +optional
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty" protobuf:"bytes,19,rep,name=envFrom"`
+	// List of environment variables to set in the container.
+	// Cannot be updated.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+```
+
+First decision:  do we need to expose both `EnvFrom` and `EnvVar` off of `Containers`, especially given that
+`EnvVar` has an optional `EnvFromSource` [ref](https://github.com/kubernetes/api/blob/v0.21.0/core/v1/types.go#L1929)
+
+```go
+// EnvVar represents an environment variable present in a Container.
+type EnvVar struct {
+	// Name of the environment variable. Must be a C_IDENTIFIER.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Optional: no more than one of the following may be specified.
+
+	// Variable references $(VAR_NAME) are expanded
+	// using the previous defined environment variables in the container and
+	// any service environment variables. If a variable cannot be resolved,
+	// the reference in the input string will be unchanged. The $(VAR_NAME)
+	// syntax can be escaped with a double $$, ie: $$(VAR_NAME). Escaped
+	// references will never be expanded, regardless of whether the variable
+	// exists or not.
+	// Defaults to "".
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	// Source for the environment variable's value. Cannot be used if value is not empty.
+	// +optional
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.
+
+Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
+`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  
+
+With that, a new Shipwright type would be:
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
+   // after any Parameter variable substitution.
+   // +optional
+   DefaultValue  string
+}
+```
+
+Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 
```

Comment:
In the BuildRun, there would not anymore be a need for a default value. At the latest at this point, the value needs to be specified by the user.

Reply:
reusing the type across the board is attractive to me here

a bit of a stretch, but if a user was uncertain about the contents of a configmap or secret used with valueFrom (perhaps they reconcile as blank) having the default could a safe fallback

so I'm inclined to keep as is ... but if that reasoning is not sufficient for you, please elaborate and we'll go from there

Reply:
So a lot of good discourse to unpack there @SaschaSchwarze0.  Bottom line, I believe I need to make some adjustments.  

But how to convey those so we can reach consensus?  Here we go ... but while I was finishing this post, @SaschaSchwarze0 and I chatted on slack and also agreed to set up a video conf for us to discuss.
So while we can all continue to talk here (I also have comments from @otaviof I need to get to), this can also serve as additional input into the conversation he and I will have.

1) Overall, the notion of conveying a staging plan wrt this EP is a possibility.  I think we could incrementally add to an env var API as part of Shipwright if need be.  I'll circle back to that based on where we land.  Then, if need be, a stage 1 could be simple k/v pairing, a stage 2 maybe could be say `valueFrom` (though I think it should be a stage 1 item, with possible pull out if something unexpected arises), followed by a probable `no` for `defaults`.  More on defaults in a bit, but perhaps I've overstretched wrt that.

2) IMO, we do have uses cases for `valueFrom`, again with the caveat that we do not have to necessarily address them in an "initial" stage if when we get to implementation there is uncertainty on the nature of the change.

Those use cases?

First, I already have a link for this in this EP, but Tekton's specific example at https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#using-a-secret-as-an-environment-source where they seed the `GITHUB_TOKEN` based on what is pulled from a k8s secret, is pervasive enough IMO that I at least could see Shipwright users liking this form.

A lot of tools like git, maven, etc. used in building are big on environment variables, and in the case of sensitive data, storing that data and retrieving it from k8s Secrets is a de facto mode of operator at this point.

And a first blush,  it seems like the reconciliation of the `BuildRun` to `TaskRun` is straight forward, as we map the analogous `BuildRun` API to the `TaskRun` API seen in that link.

Hopefully with that, along with cleaning things up around the strategy already has steps/containers point you made (more on that below), that will remove the complexity concern.

But I think it is fair to say that a path to Shipwright will be Tekton users that desire the value in standardizing image building.  And even though Tekton is a Shipwright implementation detail, analogous semantics like this, that leverage core k8s resources, seems like a distinct possibility.

Second, we also have this analogous scenario wrt build tools and this precise feature with OpenShift Builds.  Please see https://docs.openshift.com/container-platform/4.7/cicd/builds/creating-build-inputs.html#builds-using-secrets-as-environment-variables_creating-build-inputs

And from a historical perspective, that feature arose from customer feedback (again, how they are use to supplying tokens to build tools via env vars).  Long term, as we migrate OpenShift Build users to Shipwright, not having this could be a feature gap and obstruction to migration.  

Now, back to the staging, again, perhaps we do not do this initially based on how the initial implementation goes, but I think we should strive for doing it initially, and only back it out if we see encounter something unexpected in mapping from BuildRun to TaskRun during the implementation PR.

But minimally though, if we don't have direct mapping of this features for these two potential user sets, we then have a more complex migration of "how you do this in a different form via Shipwright" story that we'll need to convey.

3) Defaulting ... yeah, my motivation here stemmed from my exploration of Tekton Triggers, and their use of defaults with the parameters for `TriggerTemplates`.  Fixed some bugs in this space, and overall the feature resonated with me.

That said, none of the existing stakes in the ground wrt environment variables I noted above applied the defaulting concept.

Based on that and your initial interpretation, I'm going to back off of it in the main proposal section.  I'll probably add some mention of it in alternatives, where I'll mention the idea was entertained, but we moved off of it.

4) How it comes together .... strategy already having steps/containers.  Yeah I missed the need for more detail here, thanks.  And I think your comment hit on the choice we have here.

Setting env var's on the build step(s) is already available, so is that sufficient for setting things at the `BuildStrategy` level?  
Do we allow setting of env vars at the `BuildStrategySpec` level, that apply to all steps/containers?

I could see staging coming into play here, where we at least initially we do not add a new field for multi container override, and then wait for user feedback.

Then, separate from those choices, the precedence decision I've noted in the doc, where the admin and dev sort out whether the `BuildStrategy` or `BuildRun` take precedence, or if we do not provide that choice and decide on an order, at least initially builds on top of those choices above.

Controlling the precedence order could also be another staging item.  Where initially we just pick an order, perhaps "developer" focused, where the BuildRun wins, and then we let administrators block that with kyverno / gatekeeperm though I could go the other way as well, and then add configuration / control around the order later.